src/components/Modal.tsx

```diff
@@ -0,0 +1,84 @@
+import { FC, PropsWithChildren, createContext, useContext, useState } from 'react';
+import { createPortal } from 'react-dom';
+
+import classes from './Modal.module.css';
+import { useOutsideClick } from '../hooks/useOutsideClick';
+
+interface ModalContextType {
```

Comment:
The overall solution looks overcomplicated. This either needs some better code splitting or let's re-evaluate solution itself.